cql3/restrictions/statement_restrictions.cc

```diff
@@ -1178,6 +1178,45 @@ bool is_satisfied_by(
         }, restr);
 }
 
+bytes_opt get_bound(const expression& restr, const query_options& options, statements::bound bnd) {
+    return std::visit(overloaded_functor{
+            [&] (const conjunction& conj) {
+                if (!conj.children.empty()) {
+                    auto invoke_get_bound = [&] (const ::shared_ptr<expression>& p) {
+                        return get_bound(*p, options, bnd);
+                    };
+                    // For now, assume conjunction elements all have the same LHS.
+                    std::vector<bytes_opt> children_bounds(
+                            boost::make_transform_iterator(conj.children.cbegin(), invoke_get_bound),
+                            boost::make_transform_iterator(conj.children.cend(), invoke_get_bound));
+                    if (is_start(bnd)) {
+                        return *boost::max_element(children_bounds);
+                    } else {
+                        return *boost::min_element(children_bounds, [] (const bytes_opt& a, const bytes_opt& b) {
+                            // Default comparator ranks nullopt lower than any value; we want the opposite here.
+                            return (a && b) ? a < b : a.has_value();
```

Comment:
Actually, if we have any nullopt in there, we should return false (regardless of bnd).

"a < 3 AND a < NULL" is always false, because "a OP NULL" is always false.

Reply:
And, we must use the type comparator, not bitwise comparator.

Reply:
> Actually, if we have any nullopt in there, we should return false (regardless of bnd).

Suppose we're evaluating `col < 2 AND col < 4 AND col > 0`.  The upper bound should clearly be 2 -- the min_element of [2, 4, null].  But if `2 < null` evaluates to false, as you suggest, then null will be returned as the upper bound, which is wrong.

Reply:
> And, we must use the type comparator, not bitwise comparator.

I have to think about this, as there is no type for multi-column restrictions.

Reply:
null is correct as the upper bound in your example, because is poisons the following calculations made with it. But it's likely the caller isn't prepared for it.

It's a bit like NaN in floating point, whatever it touches it poisons.

And btw `2 < NULL` evaluates to `NULL`, not false. When checked in a WHERE clause, the `NULL` is treated as false.

Probably we should short-circuit the whole thing if NULL if present.

btw it is likely the existing code is broken here.

Reply:
> null is correct as the upper bound in your example,

No, it is not.  Recall that we're not talking about `is_satisfied_by` here but `get_bound`.  This result is never seen by the user; it is used to calculate the smallest slice to fetch from the data layer.  NULL means unbounded in this scenario, ie, fetch all rows.  The correct upper bound is 2, as we don't need to fetch any rows with values higher than 2.

Reply:
Suppose the query is `SELECT * FROM foo WHERE pk = :pk AND ck < :bound1 AND ck < :bound2`. Suppose the user set :bound2 = NULL. Then we don't want to fetch all rows, but no rows, since ck < NULL is always false for any ck.

You also might also need return something for unbounded (if there was no limit in the query, but that is distinct. It can also be done during prepare time.

Reply:
> > And, we must use the type comparator, not bitwise comparator.
> 
> I have to think about this, as there is no type for multi-column restrictions.

Each column has its own type. See the 9-input lexicographical_tri_compare().

Reply:
> > Suppose the query is `SELECT * FROM foo WHERE pk = :pk AND ck < :bound1 AND ck < :bound2`. Suppose the user set :bound2 = NULL. Then we don't want to fetch all rows, but no rows, since ck < NULL is always false for any ck.
> 
> In practice, this is what comes back:
> 
> ```
> fatal error: in "regular_col_slice": query 'select ck from tb where pk = 1 and ck < null' failed: exceptions::invalid_request_exception (Invalid null value for clustering key part SLICE(> null, < null))
> ```

Ok, so we chicken out. But for non-key columns we're broken:

```
cqlsh:keyspace1> select * from t1 where v > NULL allow filtering;

 pk | ck | v
----+----+----
  1 |  1 |  1
  1 |  2 |  2
  2 |  1 | 21
  2 |  2 | 22
```

Same for `v < NULL`.

> 
> More importantly, the existing `bounds()` function treats `col < null` as always true for regular columns.
> 

That's a bug. Converting code rather than rewriting it also carries a risk, that of copying the existing code's bugs. But at least they aren't regressions.

> > You also might also need return something for unbounded (if there was no limit in the query,
> 
> What would that be? Maybe the type's max/min?

Not all types have max/min (what's the maximum blob?). We can return a variant<unbounded, bytes_view_opt> (where a disengafed bytes_view_opt indicates a NULL bound).

Another option is optional<bytes_view_opt>, where the outer optional indicates whether the bound exists, and the inner one indicates if it's NULL or a value.

But let's start with 1:1 duplication of existing behabior, bugs and all, and follow up with a bug sweep later.

Reply:
Looks like Cassandra 3.11 is similarly broken, according to the NULL rules you mention (citation needed, btw...):
```cql
cqlsh:ks> SELECT * FROM t WHERE pk=1 AND ck>null;

 pk | ck | r
----+----+----
  1 | 11 | 21

(1 rows)
cqlsh:ks> SELECT * FROM t WHERE ck>null ALLOW FILTERING ;

 pk | ck | r
----+----+------
  1 | 11 |   21
  2 | 12 | null

(2 rows)
cqlsh:ks> SELECT * FROM t WHERE ck<null ALLOW FILTERING ;

 pk | ck | r
----+----+---

(0 rows)
```
In the mirror-image of our limitations, Cassandra allows `< null` on clustering key but not on primary key or non-key columns.

Also, Cassandra disallows null-equality restrictions `c=null` and `m[k]=null`.  We allow both, but they yield false and true, respectively, for all values.

:man_shrugging:

Reply:
Let's disallow both and then reconsider. But not as part of this change, either before or after.

Reply:
>(citation needed, btw...):

https://en.wikipedia.org/wiki/Null_(SQL)#Comparisons_with_NULL_and_the_three-valued_logic_(3VL)

Reply:
Here's what postgresql thinks about the matter:

```
postgres=# select 3 > 2 as x;
 x 
---
 t
(1 row)

postgres=# select 3 > NULL as x;
 x 
---
 
(1 row)
```

Reply:
And:

```
postgres=# select (3 > NULL) IS NULL as x;
 x 
---
 t
(1 row)
```

Reply:
> Let's disallow both and then reconsider. But not as part of this change, either before or after.

I presume you mean both `=null` cases,`<null` restrictions on different column types?

Reply:
> > (citation needed, btw...):
> 
> https://en.wikipedia.org/wiki/Null_(SQL)#Comparisons_with_NULL_and_the_three-valued_logic_(3VL)

Any citations from the language we are implementing here? ;)

Reply:
> > > (citation needed, btw...):
> > 
> > 
> > https://en.wikipedia.org/wiki/Null_(SQL)#Comparisons_with_NULL_and_the_three-valued_logic_(3VL)
> 
> Any citations from the language we are implementing here? ;)

CQL is under-specified. The goal is to be similar to SQL. Having slightly different semantics is a sure way to trap users.

So, in the absence of a specification, I fall back to SQL. This includes edge cases which could have happened in Cassandra by accident.

Reply:
> > Let's disallow both and then reconsider. But not as part of this change, either before or after.
> 
> I presume you mean both `=null` cases, not `<null` restrictions on different column types?

I mean that if Casasndra disallows something, let's disallow it too for now instead of trying to fix it now.

If it allows something we can see if is consistent with SQL or not, and if not, think about what to do.

Reply:
Now using type comparator.  Turns out that this is never invoked for multi-column restrictions, so no need to worry about that comparison.

Added `bound_t` to capture the comparator type, with a special state for unbounded columns.

Disallowed comparisons that Cassandra disallows in a4a5b77.